toolchain/sem_ir/block_value_store.h

```diff
@@ -82,6 +102,37 @@ class BlockValueStore : public Yaml::Printable<BlockValueStore<IdT>> {
   }
 
  private:
+  // A canonical block, for which we allocate a deduplicated ID.
+  struct CanonicalBlock {
+    // This is mutable so we can repoint it at the allocated data if insertion
+    // succeeds.
+    mutable llvm::ArrayRef<ElementType> data;
+
+    friend auto CarbonHashValue(CanonicalBlock block, uint64_t seed)
+        -> HashCode {
+      Hasher hasher(seed);
+      hasher.HashSizedBytes(block.data);
+      return static_cast<HashCode>(hasher);
+    }
+  };
+
+  struct CanonicalBlockDenseMapInfo {
```

Comment:
You seem to be doing some nuanced work here with SpecialData, would it be worth some comments explaining how this is intended to work?

Reply:
Added.